Java concurrency, support for synchronization

Degiorgio opened the pull request:
PR enables basic support for  analyzing concurrent java programs. Specifically, it adds support for synchronized methods,  synchronized blocks and multiple regressions tests. 

test-gen PR:  https://github.com/diffblue/test-gen/pull/1908

dcattaruzza committed:
JBMC: Modified the instrumentation of monitorexit/enter instructions

The monitorenter and monitorexit instructions are used by the JVM to
coordinate access to an object in the context of multiple threads.

We have previously added two methods to the object model that use a
counter to implement a reentrant lock. Calls to
'org.cprover.CProver.atomicBegin:()V"' and
'org.cprover.CProver.atomicEnd:()V' ensure that multiple threads do
not race in the access/modification of this counter.

In-order to support synchronization blocks, when the
monitorexit/moniitorenter bytecode instruction is executed JBMC must
call the aforementioned object model. To this end,  this commit makes
the following changes:

1. Transforms the monitorenter and monitorexit bytecode instructions
   into function-calls to the object model. Specifically,
   'java.lang.Object.monitorenter:(Ljava/lang/Object;)V' and
   'java.lang.Object.monitorexit:(Ljava/lang/Object;)V'.

2.  Transforms 'org.cprover.CProver.atomicBegin:()V"' and
    'org.cprover.CProver.atomicEnd:()V' into the appropriate
    codet instructions.

Added the appropriate target-handlers if monitorenter or monitorexit
are in the context of a try-catch block.

Degiorgio committed:
JBMC: Zero-initialized 'cproverMonitorCount' component and removed
'@lock' field (fixes #2307)

The 'cproverMonitorCount' field is a counter in the 'java.lang.Object'
model (part of the java core models library). This field is used to
implement a critical section and is thus necessary to support
concurrency.

This commit makes sure that this field (if present) is always zero
initialized as it is not meant to be non-deterministic.

This field is present only if the java core models library is loaded.

Additionally, the commit removes '@lock' field from root class
(usually: 'java.lang.Object') as it has been superseded by a locking
mechanism implemented in the java core models library.

Modified relevant unit/regression tests to reflect this change.

Degiorgio committed:
JBMC: Regression tests for multi-threaded java programs

Degiorgio committed:
JBMC: Support for synchronized methods

This commit adds support for synchronized methods by instrumenting all
methods marked with the synchronization flag with calls to
'monitorenter' and 'monitorexit'. These two methods are located in the
java models library and implement a critical section.

To this end the following changes are made:

1. New irep_id, 'is_synchronized', to represent the synchronized keyword
   and appropriate changes to 'java_byecode_convert_method.cpp' to set
   this flag when a synchronized method is encountered.
2. Setting of the 'is_static' flag when the method in question is static.
3. Functions to find and instrument synchronized methods. Specifically,
   calls to "java::java.lang.Object.monitorenter:(Ljava/lang/Object;)V"
   and "java::java.lang.Object.monitorexit:(Ljava/lang/Object;)V" are
   respectively instrumented at the start and end of a synchronized
   method . Note, the former is instrumented at every point where
   the execution of the synchronized methods terminates. Specifically
   out of order return statements and exceptions.

Static synchronized methods are not supported yet as the synchronization
semantics for static methods is different (locking on the class instead
the of the object). Upon encountering a static synchronized method the
current implementation will ignore the synchronized flag.  (showing a
warning in the process). This may obviously result in superfluous
interleavings.

Note: instrumentation of synchronized methods is only triggered if the
      '--java-threading' command line option is specified.

Note': instrumentation of synchronized methods requires the use of the
       java core models library as the locking mechanism is implemented
       in the model 'java.long.Object'.

Degiorgio committed:
JBMC: Regression tests for synchronized methods

tautschnig commented on jbmc/regression/jbmc-concurrency/get-current-thread/test2.desc:
```diff
@@ -0,0 +1,7 @@
+KNOWNBUG
```
Please add a comment that explains what the problem is.

tautschnig commented on jbmc/regression/jbmc-concurrency/get-current-thread/test3.desc:
```diff
@@ -0,0 +1,7 @@
+KNOWNBUG
```
As above.

Degiorgio replied:
done.  test2.desc and test3.desc do not work due to the same bug. I created another  test 'get-current-thread/test_bug.desc' that showcases the issue.   I also added comment over there explaining what the issue is exactly.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+/// /param name: base name of the temporary variable, to be append with "_tmp"
+/// /param type: type of the temporary variable
+/// /return returns new variable
+static symbol_exprt tmp_variable(
```
Use `get_fresh_aux_symbol`.

Degiorgio replied:
done

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+
+  // If the functions for monitorenter or monitorexit do not exist
+  // (this can only happen if the java-models libary was not downloaded).
+  // We implement them as skips because symex would crash, since it cannot find
```
Any such "crash" (whatever that is) would be a bug and that needs to be fixed. I'm confused.

Degiorgio replied:
This is badly worded. If the methods "java::java.lang.Object.monitorenter:(Ljava/lang/Object;)V" and "java::java.lang.Object.monitorexit:(Ljava/lang/Object;)V"  are not found (this happens  when the java-core-models-library is  not loaded and the method in question is synchronized).  Symex will rightfully complain as it cannot find the symbols associated with function calls to the aforementioned methods (Note: we prevent JBMC from stubbing these methods). To avoid this and thereby ensuring JBMC works with and without  the models,  we simply replace such calls with skips when the  java-core -models-library  is not loaded.

Degiorgio replied:
Will reword comment to make this clear.

kroening replied:
I'll do a PR that adds these symbols.

Degiorgio replied:
@kroening In your PR (https://github.com/diffblue/cbmc/pull/2286/) you are unconditionally adding the symbols for monitorenter/exit, while what's needed is adding them *only* when core models cannot be found (or simply returning code_skipt, as this pr is currently doing). Note: that the actual locking mechanism is implemented in these models as such they are not  empty stubs.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+
+  // Otherwise we create a function call
+  code_function_callt call;
+  call.function() = symbol_exprt(symbol, type);
```
That `type` must be the same that the symbol in the symbol table has, but here you are constructing it from scratch. This is at best redundant.

Degiorgio replied:
Your right ofc, done.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+  const irep_idt &statement=code.get_statement();
+  if(statement == ID_return)
+  {
+    // Relace the return with a monitor exit plus return block
```
Typo: s/Relace/Replace/

Degiorgio replied:
done

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+  }
+  else if((statement == ID_label)
+    || (statement == ID_block)
+    || (statement == ID_decl_block))
```
Unnecessary parentheses

Degiorgio replied:
done

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+    // Create a unique catch label and empty throw type (i.e any)
+    // and catch-push them at the beginning of the code (i.e begin try)
+    code_push_catcht catch_push;
+    irep_idt handler("pc_synchronized_catch");
```
Is this a magic string?

Degiorgio replied:
Yes, it  is used to create a unique catch label. It only used locally in that function right now,  should it be prefixed with 'CPROVER'?

tautschnig replied:
How would you guarantee that it genuinely is unique? I'd recommend to use some string that cannot possibly be used by a user, for example by using a `-` somewhere in there.

Degiorgio replied:
done.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -361,3 +526,59 @@ void convert_threadblock(symbol_tablet &symbol_table)
+/// \param symbol_table: a symbol table
+void convert_synchronized_methods(symbol_tablet &symbol_table)
+{
+  for(auto entry : symbol_table)
```
`const auto &entry`

Degiorgio replied:
done

tautschnig reviewed (commented):
Some of my comments I'd normally consider blockers, but really I'm not authoritative on Java.

To help reviewers, please please please be consistent in your code formatting (with the clang-format prescribed one being the preferred style). It's quite distracting as-is.

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+  const typet &type)
+{
+  irep_idt base_name = name + "_tmp";
+  irep_idt identifier = prefix + "::" + id2string(base_name);
```
const

Degiorgio replied:
addressing by using 'get_fresh_aux_symbol' instead.

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+/// \param symbol_table: a symbol table
+/// \param is_enter: Indicates whether we are creating a monitorenter or exit.
+/// \param object: An expression representing object  the object that need to
+/// perform a monitorenter/monitorexit operation
```
there is some mistake in this sentence

Degiorgio replied:
yep, fixed

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+/// \param object: An expression representing object  the object that need to
+/// perform a monitorenter/monitorexit operation
+static codet get_monitor_call(
+  symbol_tablet &symbol_table,
```
Is the symbol_table modified? if no it should be const and if yes it should be documented

Degiorgio replied:
no it should be constant. fixing.

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+  if(is_enter)
+    symbol="java::java.lang.Object.monitorenter:(Ljava/lang/Object;)V";
+  else
+    symbol="java::java.lang.Object.monitorexit:(Ljava/lang/Object;)V";
```
use ternary if

Degiorgio replied:
done

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+    symbol="java::java.lang.Object.monitorexit:(Ljava/lang/Object;)V";
+
+  symbol_tablet::symbolst::const_iterator it
+    = symbol_table.symbols.find(symbol);
```
use auto

Degiorgio replied:
done

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +89,171 @@ static const std::string get_thread_block_identifier(
+}
+
+/// Introduces a monitorexit before every return recursively
+/// \param code current element to check
```
not really checking but rather modifying

Degiorgio commented:
@tautschnig, @romainbrenguier  addressed the issues you raised.

thk123 commented:
@Degiorgio please put `fixes #2307` in a commit message as part of this PR (and preferably add the example as a regression if not already covered by an existing test).

cesaro commented on jbmc/src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -222,10 +222,13 @@ void java_simple_method_stubst::create_method_stub(symbolt &symbol)
-    sym.name != "java::monitorenter" && sym.name != "java::monitorexit")
+  if(!sym.is_type && sym.value.id() == ID_nil &&
+    sym.type.id() == ID_code &&
+    // do not stub internal locking calls
```
Explain why they are excluded from being stubbed

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1077,6 +1077,18 @@ void java_object_factoryt::gen_nondet_struct_init(
+      // Zero-initialize 'monitorCount' field as it is not meant to be nondet.
+      // This field is only present when the java core models are embedded. It
+      // is used to implement a critical section, which is necessary to support
+      // concurrency.
```
The field `monitorCount` in our model of `java.lang.Object` has to be renamed, as otherwise it may conflict with a field with equal name in the code under analysis. I propose `cproverMonitorCount`.

cesaro replied:
This modification has been implemented in your 3rd commit. It should be merged to the first commit. Don't forget to modify the commit message on thee 1st commit as well.

cesaro replied:
done

cesaro commented on jbmc/src/java_bytecode/java_root_class.cpp:
```diff
@@ -75,4 +75,15 @@ void java_root_class_init(
+  // This field is only present when the java core models are embedded. It is
+  // used to implement a critical section, which is necessary to support
+  // concurrency.
+  if(root_type.has_component("monitorCount"))
```
Update name, as suggested above.

Degiorgio replied:
done

cesaro commented on jbmc/regression/jbmc-concurrency/synchronized-blocks/test.desc:
```diff
@@ -0,0 +1,9 @@
+^VERIFICATION SUCCESSFUL$
+--
+--
+tests that synchronization blocks do not cause issues when the java core models library is not loaded
```
Capitalize the first word, as in previous `.desc` files.

Degiorgio replied:
done

cesaro reviewed (changes requested):
Review of the first 4 commits. We need to change the name of the monitor count, and other smaller changes.

cesaro commented on jbmc/src/java_bytecode/simple_method_stubbing.cpp:
```diff
@@ -222,10 +222,15 @@ void java_simple_method_stubst::create_method_stub(symbolt &symbol)
+    sym.type.id() == ID_code &&
+    // do not stub internal locking calls as 'create_method_stub'
+    // does not automatically create the appropriate symbols for
+    // the formal parameters.
```
This is only half of the reason. Add `, and symex will crash when a function has no formal parameters`, or whatever the reason is.

Degiorgio replied:
done

cesaro reviewed (commented):
Merge 3rd commit into 1st

cesaro commented on jbmc/src/java_bytecode/Makefile:
```diff
@@ -8,7 +8,7 @@ SRC = bytecode_info.cpp \
       java_bytecode_convert_class.cpp \
       java_bytecode_convert_method.cpp \
-      java_bytecode_convert_threadblock.cpp \
+      java_bytecode_concurrency_instrumentation.cpp \
```
In the message of previous-last commit, please explain what the behavior of JBMC will be in the presence of a static synchronized method.

Degiorgio replied:
done

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,141 @@ static const std::string get_thread_block_identifier(
+/// The resulting thread-safe codet is stored in the value field of \p symbol.
+///
+/// \param symbol_table: a symbol_table
+/// \param symbol: writeable symbol hosting code to synchronized
```
Typo: `to synchronize`

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,141 @@ static const std::string get_thread_block_identifier(
+{
+  codet &code = to_code(symbol.value);
+
+  // Get interposition code for the monitor lock/unlock
```
Library interposition is a technique very different from what's going on in here. Rephrase to say something like `Get the calls to the functions that implement the lock/unlock mechanism.`.

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,141 @@ static const std::string get_thread_block_identifier(
+/// Transforms the codet stored in in the value field of \p symbol. This
+/// expected to be a codet that needs to be synchronized, the transformation
+/// makes it thread-safe as described in the documentation of function
+/// convert_synchronized_methods
```
Describe here in pseudocode how the original `codet` is being transformed into a synchronized block, similar to how we documented the handling of class initializers.

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.h:
```diff
@@ -5,11 +5,13 @@ Module: Java Convert Thread blocks
 
 void convert_threadblock(symbol_tablet &symbol_table);
 
+void convert_synchronized_methods(symbol_tablet &symbol_table);
```
Nit: remove empty line 14?

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -361,3 +497,58 @@ void convert_threadblock(symbol_tablet &symbol_table)
   }
 }
+
+/// Iterate through the symbol table to find and appropriately instrument
```
The word `appropriately` adds no information. Remove it.

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -361,3 +497,58 @@ void convert_threadblock(symbol_tablet &symbol_table)
+/// Specifically, out of order return statements and exceptions. The latter
+/// is dealt with by instrumenting a try catch block.
+///
+/// Note: current version cannot handle static methods.
```
Describe here what the effect of this lack of support will be for those methods.

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -361,3 +497,58 @@ void convert_threadblock(symbol_tablet &symbol_table)
+/// Note': This method requires the availability of
+///   "java::java.lang.Object.monitorenter:(Ljava/lang/Object;)V" and
+///   "java::java.lang.Object.monitorexit:(Ljava/lang/Object;)V".
+///   (java-library-models)
```
Add: what will happen if those methods are not present.

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -361,3 +497,58 @@ void convert_threadblock(symbol_tablet &symbol_table)
+
+    bool is_static = (symbol.type.get(ID_is_static) == "1");
+    if(is_static)
+      // FIXME: handle static synchronized methods
```
Ok, static synchronized methods seem to be treated as simply static methods. This will produce many dataraces and unexpected results. Warn the user about this. Example: https://github.com/diffblue/cbmc/blob/develop/src/goto-instrument/remove_function.cpp#L44

Degiorgio replied:
yes as the synchronization semantics for static methods is  different (lock on the class instead the object). This is limitation with the current implementation. I have no issues with adding the warning  method, but I think it makes more sense to add it outside (where we  set the static/synchronized  irep flags)  such that we don't have to pass around the message handler reference.

cesaro replied:
Yes, makes sense.
In that case use a `PRECONDITION` in `instrument_synchronized_code` to ensure it won't be called with a static synchronized method, and do the warning in this method.

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,141 @@ static const std::string get_thread_block_identifier(
+  }
+}
+
+/// Transforms the codet stored in in the value field of \p symbol. This
```
Nit: value field of \p symbol --> field \c symbol.value

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,141 @@ static const std::string get_thread_block_identifier(
+}
+
+/// Transforms the codet stored in in the value field of \p symbol. This
+/// expected to be a codet that needs to be synchronized, the transformation
```
This expected to be ... -> The \p symbol is expected to be a Java synchronized method. Java synchronized methods do not have explicit calls to the instructions `monitorenter` and `monitorexit`, the JVM interprets the `synchronized` flag in a method and uses monitor of the object to implement locking and unlocking. Therefore JBMC has to emulate this same behavior. We insert a call to our model of `monitorenter` at the beginning of the method and calls to our model of `monitorexit` at each return instruction. We wrap the entire body of the method with an exception handler that will call our model of `monitorexit` if the method returns exceptionally.

Degiorgio replied:
done

cesaro commented on jbmc/regression/jbmc-concurrency/synchronized-method-illegal-state/test.desc:
```diff
@@ -0,0 +1,13 @@
+--
+Checks all possible paths to ensure a synchronized method does not end in an illegal monitor state.
+
+This is currently not working as explicit throws have been removed from the underlying model due to performance considerations.
```
I'm not sure this is a known bug. The code in the `.java` file should be supported without problem by this PR. What this phrase refers to is the lack of throwing a runtime exception in our model of `monitorenter/exit` *when there are reasons to throw such exceptions* (i.e, synchronizing on a null object or calling `monitorexit` when the counter is 0), but no such conditions are triggered in this test. Therefore this should work.

Am I missing something?

Degiorgio replied:
Yes this, test  actually works without the exceptions. fixing.

cesaro commented on jbmc/regression/jbmc-concurrency/synchronized-methods/A.java:
```diff
@@ -0,0 +1,49 @@
+  {
+    shared++;
+  }
+}
```
Add a new test with a `static synchronized` method and declare it as `KNOWNBUG`

Degiorgio replied:
done

cesaro reviewed (changes requested):
Review of the last 2 commits: better doxygen and one more tests.

cesaro commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,141 @@ static const std::string get_thread_block_identifier(
+/// Introduces a monitorexit before every return recursively
+/// \param [out] code: current element to modify
+/// \param monitorexit: codet to insert before the return
+static void monitor_exits(codet &code, const codet &monitorexit)
```
Add in the documentation of this function that running it will break all the sharing in the `code`, if there was any (which I believe it's unlikely at the stage where this is executed).

Degiorgio replied:
done

Degiorgio commented:
@tautschnig  can you take another look please?  Thx!

tautschnig commented:
@Degiorgio Taking a look now. Can you meanwhile engage on #2286 to clarify whether that does or does not solve the problem? You had earlier stated that it's not appropriate.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1311,6 +1313,20 @@ codet java_bytecode_convert_methodt::convert_instructions(
     }
+    // replace calls to CProver.atomicBegin
+    else if(statement == "invokestatic" &&
+            id2string(arg0.get(ID_identifier)) ==
```
I don't think `id2string` is necessary here (and neither below).

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2347,33 +2363,30 @@ codet java_bytecode_convert_methodt::convert_instructions(
-    else if(statement=="monitorexit")
+    else if(statement == "monitorenter" || statement == "monitorexit")
     {
+      std::string descriptor;
```
Nit pick: you can safely make this an `irep_idt`, though admittedly it doesn't make much of a difference.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2347,33 +2363,30 @@ codet java_bytecode_convert_methodt::convert_instructions(
       // becomes a function call
       code_typet type;
-      type.return_type()=void_typet();
+      type.return_type() = void_typet();
```
Use the 2-argument constructor of `code_typet`

peterschrammel replied:
Please address this using:
```
code_typet type(
  {code_typet::parametert(java_reference_type(void_typet()))}, void_typet());
```

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,186 @@ static const std::string get_thread_block_identifier(
+  bool is_enter,
+  const exprt &object)
+{
+  const std::string symbol =
```
Nit-pick: use `irep_idt` instead of `std::string`

Degiorgio commented:
@tautschnig  yes doing so now.

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,186 @@ static const std::string get_thread_block_identifier(
+  symbolt &symbol,
+  const exprt &sync_object)
+{
+  PRECONDITION(symbol.type.get(ID_is_static) != "1");
```
Use `!symbol.type.get_bool(ID_is_static)` instead of `.get(...) != "1"`

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -361,3 +542,75 @@ void convert_threadblock(symbol_tablet &symbol_table)
+      continue;
+    if(symbol.value.is_nil())
+      continue;
+    if(symbol.type.get(ID_is_synchronized) != "1")
```
As above: use `get_bool`

tautschnig commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -361,3 +542,75 @@ void convert_threadblock(symbol_tablet &symbol_table)
+    if(symbol.type.get(ID_is_synchronized) != "1")
+      continue;
+
+    bool is_static = (symbol.type.get(ID_is_static) == "1");
```
`const` and `get_bool`

tautschnig reviewed (approved):
Looks ok to me, but someone with proper knowledge of the Java code will have to approve for real.

mgudemann commented on jbmc/regression/jbmc-concurrency/get-current-thread/A.java:
```diff
@@ -0,0 +1,101 @@
+
+  // expected verification success
+  // --
+  // KNOWNBUG: see me_bug()
```
I don't understand that comment, at `me_bug` there is also no real description

Degiorgio replied:
added description

mgudemann commented on jbmc/regression/jbmc-concurrency/get-current-thread/test_bug.desc:
```diff
@@ -0,0 +1,19 @@
+^VERIFICATION SUCCESSFUL
+--
+--
+For some reason symex assigns 'g' to zero, even though it
```
remove `it`

Degiorgio replied:
done

Degiorgio replied:
done

mgudemann commented on jbmc/regression/jbmc-concurrency/synchronized-blocks-throw/Sync.java:
```diff
@@ -0,0 +1,37 @@
+      {
+        // Make sure the monitor is taken.
+        if (CProver.getMonitorCount(o) != 1)
+            assert false;
```
how about `assert(CProver.getMonitorCount(o) == 1)` ?

Degiorgio replied:
done

mgudemann commented on jbmc/regression/jbmc-concurrency/synchronized-blocks-throw/Sync.java:
```diff
@@ -0,0 +1,37 @@
+
+    // Make sure the monitor is free.
+    if (CProver.getMonitorCount(o) != 0)
+      assert false;
```
as above

Degiorgio replied:
done

mgudemann commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,186 @@ static const std::string get_thread_block_identifier(
+///   used to achieve object-level locking by calling monitoroenter/monitorexit.
+static codet get_monitor_call(
+  const symbol_tablet &symbol_table,
+  bool is_enter,
```
I would prefer an `enum class` for this

Degiorgio replied:
Unless you insist I will leave this as it is as there is only two possible options

mgudemann commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,186 @@ static const std::string get_thread_block_identifier(
+static void monitor_exits(codet &code, const codet &monitorexit)
+{
+  const irep_idt &statement = code.get_statement();
+  if(statement == ID_return)
```
what about `ID_throw`?

Degiorgio replied:
This is handled, by instrumenting  try { ... } catch{java.lang.throwable} {  .... }   inside every synchronized method.

mgudemann commented on jbmc/src/java_bytecode/java_bytecode_concurrency_instrumentation.cpp:
```diff
@@ -89,6 +90,186 @@ static const std::string get_thread_block_identifier(
+  codet monitorenter = get_monitor_call(symbol_table, true, sync_object);
+  codet monitorexit = get_monitor_call(symbol_table, false, sync_object);
+
+  // Create a unique catch label and empty throw type (i.e any)
```
`i.e.`

Degiorgio replied:
done

mgudemann commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2347,33 +2368,30 @@ codet java_bytecode_convert_methodt::convert_instructions(
-    else if(statement=="monitorexit")
+    else if(statement == "monitorenter" || statement == "monitorexit")
     {
+      irep_idt descriptor;
```
maybe use `... descriptor = .. ? .. : ... ;` here

Degiorgio replied:
done

mgudemann commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1065,9 +1064,17 @@ void java_object_factoryt::gen_nondet_struct_init(
     {
-      continue;
+      // Zero-initialize 'cproverMonitorCount' field as it is not meant to be nondet.
+      // This field is only present when the java core models are embedded. It
```
what does embedded mean here?

Degiorgio replied:
embedded here means that the java core models library has been loaded, i.e: in the class path. I will clarify this.

mgudemann reviewed (approved):
looks good from my side, but please add the changes that @cesaro asked for

cesaro commented:
All good from my side, many thanks @Degiorgio for your perseverance.

As for https://github.com/diffblue/cbmc/pull/2286, it's not necessary, as this PR is self contained. From the point of view of the functionality in this PR, that PR should be closed.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 711d2a4eb30f23df40cae02009dd01b737113077).
Status will be re-evaluated on next push.
Please contact peterschrammel, thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: a5e491fedcf1c6e7cc8e5569d313d4b7325a2bfc).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: f6657c127d9c0ab4c7e114ea1b41bef1430567db).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: f0aadfd1df59dc2eaff5d697a699daa9f25c100b).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: f4b455ec7f399cc80074c99b0a7cded0bf975941).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: d998ac76effcb9d6f43cc74b03247733a9beb8c6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76287768
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: d998ac76effcb9d6f43cc74b03247733a9beb8c6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76287768
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj commented:
Sorry @Degiorgio, these last 3 messages are spurious. I'm investigating.

Degiorgio committed:
JBMC: Moved format_classpath.sh to scripts/format_classpath.sh

'format_classpath.sh' is used in regression tests that make use of the
'classpath' option. This script is needed to deal with the fact that
classpath syntex is OS-dependent.

The java concurrency regression tests make heavy use of this option as
such this commit moves 'format_classpath.sh' to
'scripts/format_classpath.sh'.

Furthermore, this commit makes a very small change to 'appveyor.yml'
that enables existing java concurrency regression tests to run on
Windows.

Degiorgio commented:
@tautschnig  with regards to https://github.com/diffblue/cbmc/pull/1994 ,  I added a commit  to this PR that disables the soundness check, would appreciate your feedback.

tautschnig commented:
>  I added a commit to this PR that disables the soundness check, would appreciate your feedback.

@Degiorgio, thanks a lot for this work. Would you mind (also) putting it in a separate PR for it could be merged independently (and may spark some discussion)? I'll still comment on it in here in the meantime, though.

tautschnig commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -399,6 +399,9 @@ void jbmc_parse_optionst::get_command_line_options(optionst &options)
   }
+
+  if(cmdline.isset("java-threading"))
+    options.set_option("java-threading", true);
```
I would like the option being used by goto-symex to be more general (essentially, "java" should not appear anywhere in goto-symex as it's supposed to be a general tool). You might pick "allow-pointer-unsoundness" or the likes.

Degiorgio replied:
done, see  https://github.com/diffblue/cbmc/pull/2354

tautschnig commented on src/goto-symex/symex_assign.cpp:
```diff
@@ -232,7 +232,8 @@ void goto_symext::symex_assign_symbol(
     options.get_bool_option("simplify"),
-    constant_propagation);
+    constant_propagation,
+    options.get_bool_option("java-threading"));
```
To avoid repeated string lookups, could you please add a const member to `goto_symext` that is set in the constructor? (The same should be done for "simplify" above, but that's not something you've newly introduced.)

Degiorgio replied:
done, see  https://github.com/diffblue/cbmc/pull/2354 (will create a new pr for 'simplify' later, avoided here in-order to not sullie  #2354)

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 50d248950b42f98e671bcbd861729c99b80c9738).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76391048
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

Degiorgio commented:
@tautschnig done,   https://github.com/diffblue/cbmc/pull/2354

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 47b038369e2fc64ac68fe9e1f7b0504c0c5ff62d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76399247

Degiorgio committed:
AppVeyor fix: remove existing clones of the java models library.

This commit changes the AppVeyor configuration file such that
the java models library is freshly downloaded for every build. This
prevents JBMC from being called with an out-of-sync version of the
models library, resulting in non-deterministic behaviour.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 2602fb618d26af07dc71ed3a3c54920759b1491d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76409052

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -1634,13 +1650,26 @@ codet java_bytecode_convert_methodt::convert_instructions(
+    else if(statement == "monitorenter" || statement == "monitorexit")
     {
-      c = convert_monitorexit(i_it->source_location, op);
+      const irep_idt descriptor = (statement == "monitorenter") ?
```
Could you nonetheless factor out this block into a separate method `convert_monitorenterexit` or similar, following @romainbrenguier's model?

Degiorgio replied:
done

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: a52c9387c8dd331878e7274c6c4dd1abbceb331a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76428689
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ff79ee5f64985a8adb65452dd81adc52926d50a3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76429880

peterschrammel commented on jbmc/regression/jbmc-concurrency/java.lang.runnable/A.java:
```diff
@@ -0,0 +1,87 @@
+// JBMC, concurrency tests that make use of
```
dots within directory names are unusual here...

Degiorgio replied:
renamed

peterschrammel commented on jbmc/regression/jbmc-concurrency/synchronized-blocks-throw/Sync.java:
```diff
@@ -0,0 +1,35 @@
+      synchronized (o)
+      {
+        // Make sure the monitor is taken.
+        assert(CProver.getMonitorCount(o) == 1);
```
This will not work because the body of`CProver.getMonitorCount()` will not be part of the models library...

Degiorgio replied:
Yes, nonetheless I thought it was an important test that should be included for now, when we separate them we can either disable it or preferably  convert it to a stub and appropriately replace `CProver.getMonitorCount` in JBMC

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 2dc4dc8878adf59e663f89cfe4413d3d530e1713).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76451356

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: d42722c7fcbbc53b9b84b72a6acf4b565c6b1264).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/76943253

Degiorgio commented:
@peterschrammel https://github.com/diffblue/test-gen/pull/1908 is passing this can be merged.

cesaro commented:
Rebased to to latest `develop` after merges these last days. TG bump also rebased to latest develop. Will merge in some hours if all CIs pass.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 488bce6dc0f8b3139074c2095aa45611ac8edb68).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77131218

tautschnig commented:
In #2415 I have done some sort of emergency fix of the coreModels test. You will thus need to rebase to resolve the conflict, but I note that #2388 has yet another change proposal to this very same test. @cesaro @peterschrammel please synchronise on this.

peterschrammel commented:
@cesaro, the java models library label cbmc-5.9 should work with the current cbmc version. HEAD should work with this PR.

cesaro commented:
@peterschrammel I agree with your reasoning in the message of the commit that fixes the `coreModels` test in #2388. Now that @tautschnig has temporarily fixed the thing I will just remove my fix here.

Pointing to a specific label of the java core library is orthogonal to this PR, which as you remark, will work with both HEAD and that label in the core library repo.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: d770bb9a52990c7236eecd1aa108aaacc39b8a7c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77136906

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ce2e85e0b1403ff8058aefd2891335efb18121ad).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77147566

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 68ac56610f166bfd978ad30c0262fa15d9bf76b8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77148734
